libbeat/licenser/elastic_fetcher.go

```diff
@@ -0,0 +1,117 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package licenser
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+const xPackURL = "/_xpack"
+
+var params = map[string]string{
+	"human": "false",
+}
+
+// UnmarshalJSON takes a bytes array and convert it to the appropriate license type.
+func (t *LicenseType) UnmarshalJSON(b []byte) error {
+	if len(b) <= 2 {
+		return fmt.Errorf("invalid string for license type, received: '%s'", string(b))
+	}
+	s := string(b[1 : len(b)-1])
+	license, ok := licenseLookup[s]
+	if ok {
+		*t = license
+		return nil
+	}
+
+	return fmt.Errorf("unknown license type, received: '%s'", s)
+}
+
+// UnmarshalJSON takes a bytes array and convert it to the appropriate state.
+func (st *State) UnmarshalJSON(b []byte) error {
+	// we are only interested in the content between the quotes.
+	if len(b) <= 2 {
+		return fmt.Errorf("invalid string for state, received: '%s'", string(b))
+	}
+
+	s := string(b[1 : len(b)-1])
+	state, ok := stateLookup[s]
+	if ok {
+		*st = state
+		return nil
+	}
+	return fmt.Errorf("unknown state, received: '%s'", s)
+}
+
+// ElasticFetcher wraps an elasticsearch clients to retrieve licensing information
+// on a specific cluster.
+type ElasticFetcher struct {
+	client *elasticsearch.Client
+}
+
+// NewElasticFetcher creates a new Elastic Fetcher
+func NewElasticFetcher(client *elasticsearch.Client) *ElasticFetcher {
+	return &ElasticFetcher{client: client}
+}
+
+// Fetch connect to the cluster and call the xpack api parses the data and return an Info object.
+func (f *ElasticFetcher) Fetch() (*License, error) {
+	status, body, err := f.client.Request("GET", xPackURL, "", params, nil)
+	if status != statusOK {
```

Comment:
If the license is Apache, it will return a 405 here as the endpoint does not exist.

Reply:
Thanks, I was not sure of the behavior in the OSS world :)